Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,574 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Account
+import Shared
+import Storage
+import Sync
+import SyncTelemetry
+import AuthenticationServices
+import Common
+
+// Extends NSObject so we can use timers.
+public class RustSyncManager: NSObject, SyncManager {
+    // We shouldn't live beyond our containing BrowserProfile, either in the main app
+    // or in an extension.
+    // But it's possible that we'll finish a side-effect sync after we've ditched the
+    // profile as a whole, so we hold on to our Prefs, potentially for a little while
+    // longer. This is safe as a strong reference, because there's no cycle.
+    private weak var profile: BrowserProfile?
+    private let prefs: Prefs
+    private var syncTimer: Timer?
+    private var backgrounded: Bool = true
+    private let logger: Logger
+    private let fxaDeclinedEngines = "fxa.cwts.declinedSyncEngines"
+    private var notificationCenter: NotificationProtocol
+
+    let fifteenMinutesInterval = TimeInterval(60 * 15)
+
+    public var lastSyncFinishTime: Timestamp? {
+        get {
+            return prefs.timestampForKey(PrefsKeys.KeyLastSyncFinishTime)
+        }
+
+        set(value) {
+            if let value = value {
+                prefs.setTimestamp(value,
+                                   forKey: PrefsKeys.KeyLastSyncFinishTime)
+            } else {
+                prefs.removeObjectForKey(PrefsKeys.KeyLastSyncFinishTime)
+            }
+        }
+    }
+
+    lazy var syncManagerAPI = RustSyncManagerAPI(logger: logger)
+
+    public func migrateSyncData() {
+        // The sync functions used by the old native sync manager did not return a
```

Comment:
I think I probably mislead you here before but this comment isn't quite correct.

The persisted global state is just a [list of engines which have been declined](https://github.com/mozilla/application-services/blob/main/components/sync15/src/client/state.rs#L42-L62) - there's no per-engine state, so it will not cause anything to think there was a first-sync.

The main risk of not doing this here will be that the first "real" sync will supply [enabled_changes](https://github.com/mozilla/application-services/blob/main/components/sync_manager/src/types.rs#L14) if the user happens to have enabled/disabled an engine exactly as this upgrade was happening, and because we didn't accurately know what was "declined" before we might do the wrong thing. The fact we can be 100% sure this first sync doesn't supply that arg means we can't get confused.

However, there's actually a reasonable chance that not asking for *any* engines to sync might do the right thing here too - ie, if we can somehow tell this magic we want to sync `SyncEngineSelection::Some{ engines: vec![] }` I suspect it will do what we want.

However, ISTM there's still a bit of a risk here - what if this sync fails due to, say, a transient network error? (I can't see where it's actually called though!) We are then lumping edge-cases on edge-cases so the chances of users doing a wrong thing might be small enough we don't really care. Alternatively, we could almost certainly construct a valid `persistedState` from what we know the enabled/disabled state is already (it's just a json string), which is info we already have - then we could avoid this entirely. (But is even *that* true - for all we know, the user tried to change the enabled state just before we upgraded, but that sync failed - so we can't be 100% sure the state of the UI checkboxes are an accurate representation of the current server state can we?

I wonder if another option is to simply not pass that `enabled_changes` param on a first sync based on a pref or something? It looks like that would mean this can die entirely too, and will still do the right thing after that first sync? I think that might work except for the caveat about that sync failing?

tl;dr - this needs more thought. Maybe we should start by working out exactly what might happen if that edge-case of the user changing a checkbox before the first normal sync happens and whether we are fine with that edge-case given it seems quite unlikely

Reply:
Thanks for the clarification. So stepping back a bit, and I might be missing something here, but as the comment suggests I thought the reason for performing this "migration" was to prevent the first sync with the sync manager component taking longer than normal. If the actual issue is the edge case of the user changing their local engine enablements between the time we retrieve them and when they're passed into the sync call, is this an issue that's specific to the sync manager integration? In other words, is this an issue that could happen now in Firefox iOS release (or any other platform)?

I think for now I'm going to remove this function so that resolving how we handle this situation doesn't block the whole effort. In the meantime, I'll see if I can replicate this locally. Let me know if you have any objections @mhammond.